Fix EventEngine factory method return types

drfloob opened the pull request:
Obvious issue: since Listener and DNSResolver are abstract, they cannot
be returned by value.

This change returns unique_ptr from both factory methods. Listener objects
will be stored exclusively inside a grpc_tcp_server struct, which is
itself ref counted.

The DNSResolver situation is less clear. EventEngine implementations may want to
reuse resolver instances, which could make unique_ptrs a poor fit. However, I'm
not sure it's wise to have the EventEngine API defer to implementations on
ownership of DNSResolvers, since EventEngines would then not be interchangeable.
So I propose the API return unique_ptr here as well. I considered returning a
`DNSResolver*` with documentation that indicated the caller takes ownership, but
that's effectively an unenforceable unique_ptr (of course, with the added 
uniqueness constraint).

markdroth commented:
I absolutely agree that these can't be returned by value, and that they should not be raw pointers, since those don't allow the compiler to enforce ownership.  Using `unique_ptr<>` is one option, but there's another to consider as well.

One thing to keep in mind is that these objects will actually need to be ref-counted under the hood, because they may have internal callbacks pending at the moment at which the caller gives up ownership.  The usual pattern for dealing with that is to have each callback hold a ref to the object, so that the object isn't actually destroyed immediately when the caller gives up ownership; instead, the caller shuts down the object when giving up its ref, which causes the object to cancel any pending callbacks, and then when all pending callbacks have actually been invoked and have given up their refs, *then* the object is destroyed.  Inside C-core, we typically use `InternallyRefCounted<>` for this pattern (internally, see go/grpc-c-core-ref-counted-types-update for context), but I don't think we want to expose that type publicly.

Option 1: The closest thing to `InternallyRefCounted<>` available in the C++ standard library is `shared_ptr<>`.  However, `shared_ptr<>` does not have a way to differentiate the external owner from the internal ref-holders (i.e., it's more like our `RefCounted<>` than like our `InternallyRefCounted<>`).  This means that the interfaces would need to provide a `Shutdown()` method, which the external owners would need to explicitly call before releasing their ref.  That's workable, although it is a little sub-optimal, because (a) it does not prevent having multiple external ref-holders, which we don't need to support, and (b) it seems a little cumbersome to require the caller to explicitly call `Shutdown()` when we should ideally be able to do that automatically when the caller gives up their ref.

Option 2: The other option is to use `unique_ptr<>` here, as you suggest, and make it the implementation's problem to handle the ref-counting internally.  Implementations can do this by moving the actual functionality to a nested object that is ref-counted (probably using `shared_ptr<>`) and then having an external shim like this:

```
class MyListener : public Listener {
 public:
  // Real implementation is in ListenerImpl class.
  MyListener() : impl_(std::make_shared<ListenerImpl>()) {}
  // On destruction, shut down impl.
  ~MyListener() { impl_->Shutdown(); }

  // Delegate all methods to impl.
  absl::Status Bind(const ResolvedAddress& addr) override {
    return impl_->Bind(addr);
  }
  absl::Status Start() override { return impl_->Start(); }

 private:
  std::shared_ptr<ListenerImpl> impl_;
};
```

This way, when the caller destroys the `unique_ptr<>`, it automatically shuts down the impl and gives up its ref.  Shutting down the impl will cause the impl to cancel any pending callbacks, and when those callbacks are invoked and give up their refs, then the impl will be destroyed.

This actually makes me notice that the `Listener` API has an explicit `Shutdown()` method, but the `DNSResolver` API does not.  We should probably handle shutdown of these objects the same way in both cases.  If we go with option 1, we should add a `Shutdown()` method to `DNSResolver`; if we go with option 2, we should remove the `Shutdown()` method from `Listener`.

I would lean toward option 2, since it makes the API a little simpler.  But either one is workable.

drfloob commented:
Thanks Mark, this is a tricky problem. Good catch that the DNSResolver is missing a Shutdown method, however I'm not certain the two APIs should be the same. The `CreateListener` method should return something wholly unique every time it is called, and as we've discussed, the `GetDNSResolver` method may return a shared instance - the API contract leaves that unspecified today. If we want to make the APIs the same, I think we'd want to document like 

*An EventEngine's `GetDNSResolver` method must return an object on which `Shutdown` can be called when the caller is finished with it.*

This would at least be fair warning to not return a shared object that is expected to continue operation after the caller's lifetime. @apolcyn may I get your opinion here? I'm curious whether we'd want callers of `EventEngine::GetDNSResolver` to have to call `Shutdown` on the resolver when they're done using it, or if we want to state that resolver shutdown happens as a gRPC-shutdown-time operation (for example). I'm not sure what implementation may be more natural, if the method returns a unique resolver that the caller owns, or a shared instance that the caller borrows and discards.

Regarding your proposals Mark, I don't really like Option 2 because I do not want to place that burden on the EventEngine implementer/integrator. We'd remove the explicit `Shutdown` method and add documentation that states *"The custom EventEngine is expected to create a shim that handles internal ref counting, something like ... <50 lines of code>"*. I think it's much more complicated to write that code and the documentation for it.

I do have an alternative solution for the problem of supporting a single "special owner" with multiple ref-counted "non-owners" similar to what we do with `InternallyRefCounted<>`. As I understand it, the main difference between the two roles in this situation is that the (singular) owner is expected to call `Shutdown` before releasing their ref. So how about if we return shared_ptrs as you described in Option 1, but add one responsibility to the caller.

New EventEngine API:
```
  virtual absl::StatusOr<std::shared_ptr<Listener>> CreateListener(...) = 0;
  virtual absl::StatusOr<std::shared_ptr<DNSResolver>> GetDNSResolver() = 0;
```

A custom cleanup object that calls shutdown on either type:
```
template<class Shutdownable>
class EnsureShutdown {
 public:
  EnsureShutdown(std::shared_ptr<Shutdownable> ptr) : ptr_(ptr) {/* sanity checks */}
  ~EnsureShutdown() { ptr_->Shutdown(); }
 private:
  std::shared_ptr<Shutdownable> ptr_;
};
```

The caller would then be able to do assert ownership :
```
{
  std::shared_ptr<Listener> listener = event_engine->CreateListener(...);
  EnsureShutdown cleanup(listener);
  // callbacks are scheduled, listener is copied and accrues more refs
}  // cleanup's destructor is called, which calls listener->Shutdown(), before listener is unreffed
```

This would allow us the possibility of treating the DNSResolver in the same way as we treat Listener, or using the `std::shared_ptr<DNSResolver>` as is and letting the smart pointer do its normal cleanup. But it's probably still a problem if we don't make that lifetime expectation clear in the EventEngine API itself.

A few notes:
* templates are feasible here because it's internal code. We could open source `EnsureShutdown` but it's not necessary I don't think.
* `EnsureShutdown` would work with a raw pointer as well, but taking a shared_ptr in that object ensures correct behavior even if other intermediate code in the same scope does something silly like `listener.reset()`

markdroth commented:
> it may be more natural that the `GetDNSResolver` method could return a shared instance - the API contract leaves that unspecified today. If we want to make the APIs the same, we'd want to document something like:
> 
> _An EventEngine's `GetDNSResolver` method must return an object on which `Shutdown` can be called when the caller is finished with it._
> 
> This would at least be fair warning to not return a shared object that is expected to continue operation after the caller's lifetime. @apolcyn may I get your opinion here? I'm curious whether we'd want callers of `EventEngine::GetDNSResolver` to have to call `Shutdown` on the resolver when they're done using it, or if we want to state that resolver shutdown happens as a gRPC-shutdown-time operation (for example). I'm not sure what implementation may be more natural, if the method returns a unique resolver that the caller owns, or a shared instance that the caller borrows and discards.

I don't think either of these options actually prevents an EE implementation from returning the same object to multiple calls of `GetDNSResolver()`.  This is actually the use-case for which we use `DualRefCounted<>` in C-core: multiple external ref-holds and multiple internal ref-holders.  Note that in this case, we still need a way to know when to shut down the object, but none of the external ref-holders know how many other external ref-holders there may be, which means that none of them can actually be responsible for shutting down the object.  Instead, the EE impl will need to provide its own ref-counting mechanism to determine when the object should actually be shut down.  In essence, it will need to track how many external refs to the object it has handed out via `GetDNSResolver()`, so that it will not actually shut down until `Shutdown()` has been called the same number of times.  That's true regardless of whether we use option 1 (where `Shutdown()` is invoked by the caller) or option 2 (where `Shutdown()` is invoked on the internal object by the external object's dtor).  The code needed to do this ref-counting is basically the same in both options.

Given that, I think the ability to return multiple refs to the same object is actually somewhat orthogonal to the question of what the API should be, because it can be made to work either way.  I think the only real question here is whether we want to put the burden on the caller to have to explicitly call `Shutdown()` or whether we want shutdown to happen implicitly when the caller releases the object.

> 
> Regarding your proposals Mark, I was thinking of a shim like that, but ultimately I don't really like Option 2 because I do not want to place that burden on the EventEngine implementer/integrator. 

I don't actually see that option 2 is a huge burden.  Pretty much any EE implementation will need to deal with ref-counting anyway, and this simple wrapper doesn't seem like it adds much complexity relative to that.

**I think this is the core issue here: what trade-off do we want to make between making it easier to write an EE impl and making it easier to use the API?**  My inclination is that we actually want to make it easier to use the API, because there will probably not be that many EE impls, but there will be a lot of places where the EE API is *used* -- especially since we're now planning to make this a stable, public API, which means that it will be used by applications as well as by gRPC internally.

> We'd remove the explicit `Shutdown` method and add documentation that states _"The custom EventEngine is expected to create a shim that handles internal ref counting, something like ... <50 lines of code>"_. It's much more complicated to write that code and the documentation for it.

I don't think we need to document anything about ref-counting.  That's an implementation detail for individual EE impls, and it's not something that every EE impl would need to do.  For example, in a single-threaded system (like maybe on an embedded device), there would be no need for ref-counting at all.   This is not a property of the API, so it doesn't need to be documented as part of the API.

I could see us wanting to document this if we wind up writing a separate doc that provides hints for how to write an EE impl.  But I'm skeptical that there will ever be enough EE impls written that we will need such a doc.

> 
> I do have an alternative solution for the problem of supporting a single "special owner" with multiple ref-counted "non-owners" similar to what we do with `InternallyRefCounted<>`. As I understand it, the main difference between the two roles in this situation is that the (singular) owner is expected to call `Shutdown` before releasing their ref. So how about if we return shared_ptrs as you described in Option 1, but add one responsibility to the caller.

I don't think you're actually proposing a new responsibility for the caller here.  From the perspective of the API, the caller's responsibility is exactly what I proposed in option 1: they must call `Shutdown()` before releasing their ref to the object.  I think that what you're proposing here is a technique that the caller can use to discharge that responsibility, but that isn't actually part of the API contract itself -- the API is exactly the same regardless of whether the caller manually invokes `Shutdown()` or uses the kind of RAII-like wrapper you suggest.  It's up to each caller how to do this, as long as they meet the API contract of ensuring that `Shutdown()` is actually called.

drfloob commented:
> I don't think you're actually proposing a new responsibility for the caller here. From the perspective of the API, the caller's responsibility is exactly what I proposed in option 1: they must call Shutdown() before releasing their ref to the object.

Cool, I think we agree. Your suggestion only differs from mine on whether you call it a new responsibility or not, and reasonable folks can disagree on terms there. If we're adding an instruction that states "Create this EnsureShutdown object" because that's a convention we want to suggest to manage ownership, I'd call that a new responsibility (to create the object), even if it serves the same API contract as calling shutdown on the existing object.

> what trade-off do we want to make between making it easier to write an EE impl and making it easier to use the API?

That's a good question, and it's definitely one spectrum to consider here. It sounds like gRPC wants complex ownership stories around shared DNSResolvers - one or more owners with imperfect knowledge of each other, and multiple non-owners, each with different sets of responsibilities. I don't think an EventEngine should concern itself with that, it pushes at the boundaries of the single-responsibility principle (SRP) because valid use cases of a DNSResolver may have no conception of an owner and non-owner whatsoever.

> My inclination is that we actually want to make it easier to use the API, because there will probably not be that many EE impls, but there will be a lot of places where the EE API is used

That's a totally fair argument, and it's good to know your position. I'd prefer to maintain a relatively simple EventEngine with as few responsibilities as possible, and as little knowledge of gRPC internals as possible (SRP again, some LoD), because a simpler EventEngine is an easier thing to build, debug and maintain. The EventEngine is the thing that will get implemented multiple times (albeit maybe few times), and gRPC needs to be written to work with it just once. Nonetheless my hope is that we can find clean ways that the rest of gRPC can manage the complex lifetime stories it needs around these shared objects.

I don't have a solution for this yet, and I'd like to brainstorm it further. I think something like `DualRefCounted<>` managed inside the rest of gRPC-core (not the EventEngine) could work. I'll sync up with you to clarify some usage patterns, and look into whether we could find an approach where callers effectively anoint themselves as owners where appropriate.

drfloob commented:
Thanks Mark! I've updated the PR based on our offline discussion. The unique_ptr solution makes sense, it's flexible enough to handle various EventEngine object lifetime decisions in ways that the caller does not need to understand. I've also removed the Shutdown method on the Listener.

markdroth commented on include/grpc/event_engine/event_engine.h:
```diff
@@ -166,14 +166,17 @@ class EventEngine {
   /// Factory method to create a network listener.
-  virtual absl::StatusOr<Listener> CreateListener(
+  ///
+  /// The EventEngine implementation is responsible for defining the lifteime of
```
I don't think this comment is needed.  The fact that the return type is a `unique_ptr<>` is basically self-documenting, and enforced by the compiler.

drfloob replied:
Done

markdroth commented on include/grpc/event_engine/event_engine.h:
```diff
@@ -236,7 +239,11 @@ class EventEngine {
   /// Retrieves an instance of a DNSResolver.
-  virtual absl::StatusOr<DNSResolver> GetDNSResolver() = 0;
+  ///
+  /// The EventEngine implementation is responsible for defining the lifteime of
```
Same here.

drfloob replied:
Done

markdroth reviewed (approved):
Looks great, modulo a couple of nits.  Thanks!
